### PR TITLE
Slash command gateway events

### LINKED ISSUE #8: Support Slash Command Gateway Events
The following events need to be implemented.

- `INTERACTION_CREATE`
- `APPLICATION_COMMAND_CREATE`
- `APPLICATION_COMMAND_UPDATE`
- `APPLICATION_COMMAND_DELETE`

### PR DESCRIPTION
**Pull request by swarley**
# Summary

Adds support for Interaction and ApplicationCommand events, as well as a builder for registering an ApplicationCommand on the API.

This is a WIP, and a few things are missing. Namely, docs are incomplete, no tests are written, and some logic can be cleaned up.

However, I believe this is fairly close to how I want to public facing API to resemble.

Closes #8 

---

## Added
- TODO
## Changed
- TODO

### CODE REVIEW

File: lib/discordrb/api/application.rb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+      guild_id,
+      :get,
+      "#{Discordrb::API.api_base}/applications/#{application_id}/guilds/#{guild_id}/commands",
+      Authorization: token
+    )
+  end
+
+  # Create an application command for a guild.
+  # https://discord.com/developers/docs/interactions/slash-commands#create-guild-application-command
+  def create_guild_command(token, application_id, guild_id, name, description, options = nil)
```
Comment (Daniel-Worrall) [inline lib/discordrb/api/application.rb:73]: `options = []`

File: lib/discordrb/api/webhook.rb [hunk 2]
```diff
@@ -80,4 +80,57 @@ def token_delete_webhook(webhook_token, webhook_id, reason = nil)
...
+      webhook_id,
+      :delete,
+      "#{Discordrb::API.api_base}/webhooks/#{webhook_id}/#{webhook_token}/messages/#{message_id}",
+      {}
+    )
+  end
+
+  # Respond to an interaction.
+  # https://discord.com/developers/docs/interactions/slash-commands#create-interaction-response
+  def create_interaction_response(interaction_token, interaction_id, type, tts: nil, content: nil, embeds: nil, allowed_mentions: nil, flags: nil)
```
Comment (Daniel-Worrall) [inline lib/discordrb/api/webhook.rb:111]: Can we move this to a new api/iteraction.rb file

File: lib/discordrb/application_commands/builder.rb [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      }.compact
+    end
+  end
+
+  class Builder
+    attr_reader :name
+    attr_reader :description
+    attr_reader :options
+
+    TYPES = {
```
Comment (Daniel-Worrall) [inline lib/discordrb/application_commands/builder.rb:33]: Crystal has spoilt me. My instinct was to say make this an enum

File: lib/discordrb/application_commands/builder.rb [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  end
+
+  class GroupBuilder < Builder
+    undef group
+
+    def to_h
+      super.merge(type: TYPES[:group])
+    end
+  end
+end
```
Comment (Daniel-Worrall) [inline lib/discordrb/application_commands/builder.rb:121]: \n pls. Rubocop will yell!

**Review by Daniel-Worrall — COMMENTED** [review]
Structure lgtm. Carry on

**Comment by swarley** [comment]
Sorry about the delay on this, I'm hoping to finish this up this week.

**Comment by Daniel-Worrall** [comment]
Update: This will be closed in favor of merging slash_commands branch into main once swarley is satisfied with the API entirely.

**Comment by swarley** [comment]
Closed in favor of #71

**Pull request closed without merging**